### PR TITLE
Add captions/subtitles extraction

### LINKED ISSUE #31: Add a way to get subtitles and pass them to java (it doesn't matter if they have timing)
Add a way to get subtitles and pass them to java (it doesn't matter if they have timing)

### PR DESCRIPTION
**Pull request by sealedtx**
Resolves #31. Not sure if subtitles functionality is completely done... Will appreciate if someone can review and share thoughts

### CODE REVIEW

**Comment by xibr** [comment]
hi, 
In this case, I cannot know the subtitle file type if I do not use the default download provided in the library.
I am using android download manager.

If it is possible to check the Extension of the file translation, maybe if you add something like this

```
    for (SubtitlesInfo info : subtitles) {
                if (info.getLanguage().equals("en") && !info.isAutoGenerated() && info.getExtension().equals("vtt")) {
                    String subtitleUrl = info.getUrl();
                }
            }
```
I think he misses something like this to check.
`getExtension()`

**Comment by sealedtx** [comment]
@xibr Actually it could be no format provided in url by youtube, and I don't know which format is default, for most videos I have checked it was "srv1", but I can't be sure that it is default for all. Thank you for your feedback, probably I need to think about improvement api to support external downloaders as fully as default downloader

**Comment by xibr** [comment]
Yes, I thought about it too, I think it would be a good idea.

**Comment by sealedtx** [comment]
@xibr is it better now? You can use it like:
```
for (SubtitlesInfo info : subtitles) {
   if (info.getLanguage().equals("en") && !info.isAutoGenerated()) {
      String subtitlesUrl = info.getSubtitles()
                        .formatTo(Extension.WEBVTT)
                        .getDownloadUrl();
   }
}
```

**Comment by xibr** [comment]
Yes, it now returns the link and can be easily downloaded.

`https://www.youtube.com/api/timedtext?v=9mwRQ96CuHg&asr_langs=de,en,es,fr,it,ja,ko,nl,pt,ru&caps=asr&xorp=true&hl=en&ip=0.0.0.0&ipbits=0&expire=1595725796&sparams=ip,ipbits,expire,v,asr_langs,caps,xorp&signature=BAFB43065BB7F31DD0A5675D821666F506BB9DE2.3D9C56B09D690223B85CEEB7D5CEFB4E8B1C5CD5&key=yt8&lang=en&fmt=vtt`

I've tried several videos with subtitles so far and they all work properly.

**Comment by xibr** [comment]
i have this error when try to download from this video id `y9pfCQ5qQYY`
This video has no subtitles.
```
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean com.alibaba.fastjson.JSONArray.isEmpty()' on a null object reference
        at com.github.kiulian.downloader.parser.DefaultParser.getSubtitlesInfoFromCaptions(DefaultParser.java:117)
        at com.github.kiulian.downloader.YoutubeDownloader.getVideo(YoutubeDownloader.java:71)
```

**Comment by sealedtx** [comment]
@xibr fixed and covered with tests cases when no subtitles

File: src/main/java/com/github/kiulian/downloader/model/subtitles/SubtitlesInfo.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+public class SubtitlesInfo {
+    private final String url;
+    private final String language;
+    private final boolean isAutoGenerated;
+
+    public SubtitlesInfo(String url, String language, boolean isAutoGenerated) {
+        this.url = url
+                .replaceAll("&fmt=[^&]*", "") // existing format
```
Comment (fsanaulla) [inline src/main/java/com/github/kiulian/downloader/model/subtitles/SubtitlesInfo.java:30]: It's better to compile them before

File: src/main/java/com/github/kiulian/downloader/parser/DefaultParser.java [hunk 2]
```diff
@@ -88,17 +92,74 @@ public VideoDetails getVideoDetails(JSONObject config) {
...
+        return subtitlesInfo;
+    }
+
+    @Override
+    public List<SubtitlesInfo> getSubtitlesInfo(String videoId) throws YoutubeException {
+        String xmlUrl = "https://video.google.com/timedtext?hl=en&type=list&v=" + videoId;
+
+        String subtitlesXml = extractor.loadUrl(xmlUrl);
+
+        Pattern pattern = Pattern.compile("lang_code=\"(.{2,3})\"");
```
Comment (fsanaulla) [inline src/main/java/com/github/kiulian/downloader/parser/DefaultParser.java:147]: Can we move it upper

**Comment by xibr** [comment]
I have tried some videos and it seems to be working fine.

**Pull request merged**